### PR TITLE
stop using java clock in ScopedCache

### PR DESCRIPTION
**Pull request by strokyl**

### CODE REVIEW

**Comment by strokyl** [comment]
UP

File: zio-cache/shared/src/main/scala/zio/cache/ScopedCacheImplementation.scala [hunk 1]
```diff
@@ -123,52 +124,53 @@ private[cache] class ScopedCacheImplementation[Key, Environment, Error, Value](
               ensureMapSizeNotExceeded(key) *> scoped
             case complete @ MapValue.Complete(key, _, _, _, timeToLive) =>
               trackHit()
-              if (hasExpired(timeToLive)) {
+              if (now.isAfter(timeToLive)) {
```
Comment (guizmaii) [inline zio-cache/shared/src/main/scala/zio/cache/ScopedCacheImplementation.scala]: @strokyl I think we're loosing a piece of information here. The code was super explicit before: `hasExpired(timeToLive)`

What do you think of doing this to keep the explicitness of what this code is testing:
```suggestion
              val hasExpired = now.isAfter(timeToLive)
              if (hasExpired) {
```

File: zio-cache/shared/src/main/scala/zio/cache/ScopedCacheImplementation.scala [hunk 2]
```diff
@@ -123,52 +124,53 @@ private[cache] class ScopedCacheImplementation[Key, Environment, Error, Value](
...
                 map.remove(k, value)
                 ensureMapSizeNotExceeded(key) *> complete.releaseOwner *> ZIO.succeed(get(k))
               } else {
                 ensureMapSizeNotExceeded(key).as(complete.toScoped)
               }
             case MapValue.Refreshing(promiseInProgress, complete @ MapValue.Complete(mapKey, _, _, _, ttl)) =>
               trackHit()
-              if (hasExpired(ttl)) {
+              if (now.isAfter(ttl)) {
```
Comment (guizmaii) [inline zio-cache/shared/src/main/scala/zio/cache/ScopedCacheImplementation.scala]: Same suggestion

File: zio-cache/shared/src/main/scala/zio/cache/ScopedCacheImplementation.scala [hunk 3]
```diff
@@ -123,52 +124,53 @@ private[cache] class ScopedCacheImplementation[Key, Environment, Error, Value](
...
+        value = map.putIfAbsent(k, MapValue.Pending(newKey, scoped))
+      }
+      val finalScoped = if (value eq null) {
+        ensureMapSizeNotExceeded(newKey) *> scoped
+      } else {
+        value match {
+          case MapValue.Pending(_, scopedEffect) =>
+            scopedEffect
+          case completeResult @ MapValue.Complete(_, _, _, _, ttl) =>
+            if (now.isAfter(ttl)) {
```
Comment (guizmaii) [inline zio-cache/shared/src/main/scala/zio/cache/ScopedCacheImplementation.scala]: Sam suggestion

**Comment by guizmaii** [comment]
@adamgfraser Could you please have a look at this PR, please?

**Review by adamgfraser — CHANGES_REQUESTED** [review]
We should use the `unsafe` operators on `Clock` here so we retain the benefits of not having to evaluate a ZIO workflow to get the time while having the testability of being able to substitute our own implementation.

**Comment by strokyl** [comment]
> We should use the `unsafe` operators on `Clock` here so we retain the benefits of not having to evaluate a ZIO workflow to get the time while having the testability of being able to substitute our own implementation.

Should be good now.

**Review by adamgfraser — CHANGES_REQUESTED** [review]
Would just provide the unsafe capability explicitly for internal code like this.

```scala
hasExpired(ttl)(Unsafe.unsafe)
```

instead of:

```scala
Unsafe.unsafe(hasExpired(ttl)(_))
```

Otherwise looks great!

**Comment by strokyl** [comment]
> Would just provide the unsafe capability explicitly for internal code like this.
> 
> ```scala
> hasExpired(ttl)(Unsafe.unsafe)
> ```
> 
> instead of:
> 
> ```scala
> Unsafe.unsafe(hasExpired(ttl)(_))
> ```
> 
> Otherwise looks great!

Should be good now.

**Review by adamgfraser — APPROVED** [review]

**Pull request merged**